framework/math/quadratures/angular/triangle_quadrature.cc

```diff
@@ -0,0 +1,476 @@
+#include "framework/math/quadratures/angular/legendre_poly/legendrepoly.h"
+#include "framework/math/math.h"
+#include "framework/logging/log.h"
+#include "framework/runtime.h"
+#include <algorithm>
+#include <cmath>
+#include <iomanip>
+#include <numeric>
+
+namespace opensn
+{
+
+TriangleQuadrature::TriangleQuadrature(unsigned int method, unsigned int sn, unsigned int moments)
+  : method_(method), sn_(sn), moments_(moments)
+{
+  TriangleInit();
+}
+
+void
+TriangleQuadrature::TriangleInit()
+{
+  log.Log0() << "Method: " << method_ << "\nsn " << sn_;
+
+  if (method_ != 1 and method_ != 2 and method_ != 3 and method_ != 0)
+  {
+    log.Log0Error() << "TriangleQuarature: " << method_ << " is not a valid method.\n"
+                    << "Method must be 0, 1, 2, or 3." << std::endl;
+    Exit(510);
+  }
+  if (moments_ < 0 or moments_ > sn_)
+  {
+    log.Log0Error() << "TriangleQuarature: " << moments_ << " must be > 0 and < " << sn_
+                    << std::endl;
+    Exit(510);
+  }
+
+  // Get the gauss points for the z axis. The number of points for GL is twice
+  // that of the quadrature.
+  Vector3 new_omega;
+  const auto GAUSS_LEG_QUAD = GaussLegendreQuadrature(sn_);
+
+  // Formulate the triangular quadrature
+  VecDbl newZi, newWeights;
+  newZi.reserve(GAUSS_LEG_QUAD.qpoints.size());
+  newWeights.reserve(GAUSS_LEG_QUAD.qpoints.size());
+
+  for (size_t pos = 0; pos < GAUSS_LEG_QUAD.qpoints.size(); ++pos)
+  {
+    if (GAUSS_LEG_QUAD.qpoints[pos].x < 0)
+      continue;
+    newZi.push_back(GAUSS_LEG_QUAD.qpoints[pos].x);
+    newWeights.push_back(GAUSS_LEG_QUAD.weights[pos]);
+  }
+
+  int num_div = 1;
+  int weightPos = 0;
+  for (const auto& QUAD_POINT : GAUSS_LEG_QUAD.qpoints)
+  {
+    double deltaVPhi = M_PI / (2.0 * static_cast<double>(num_div));
+    // When the GaussLegendre quadrature gives us the x points we will
+    // use for our z positions on the unit sphere, they are ordered in
+    // descending order - largest magnitude towards 0 on the negative side,
+    // and ascending order on the positive side of 0
+    // The weights_ are defined using the weights_ given by the GL quadrature
+    // which is the position in the weights_ that weightPos keeps track of.
+    // This will ignore the positive x values, and just use the descending
+    // order given by the quadrature and use the absolute value of the x values.
+
+    if (QUAD_POINT.x >= 0.0)
+      break;
+
+    for (int v = 0; v < num_div; ++v)
+    {
+      double new_z_value = abs(QUAD_POINT.x);
+      double phi = deltaVPhi / 2.0 + (double)v * deltaVPhi;
+      double theta = acos(new_z_value);
+      double sinTheta = sqrt(1 - new_z_value * new_z_value);
+      //      double weightCurrent = GAUSS_LEG_QUAD.weights[weightPos] / (num_div);
+
+      new_omega.x = sinTheta * cos(phi);
+      new_omega.y = sinTheta * sin(phi);
+      new_omega.z = new_z_value;
+      weights_.push_back(GAUSS_LEG_QUAD.weights[weightPos] * (M_PI / (2.0 * num_div)));
```

Comment:
Since you know the number of weights, omegas, and abscissae, reserve their size before the loop.